src/librbd/librbd.cc

```diff
@@ -6140,36 +6140,55 @@ extern "C" int rbd_aio_write2(rbd_image_t image, uint64_t off, size_t len,
   return 0;
 }
 
-extern "C" int rbd_aio_writev(rbd_image_t image, const struct iovec *iov,
-                              int iovcnt, uint64_t off, rbd_completion_t c)
+static int get_iovec_length(const struct iovec *iov, int iovcnt, ssize_t &len)
 {
-  librbd::ImageCtx *ictx = (librbd::ImageCtx *)image;
-  librbd::RBD::AioCompletion *comp = (librbd::RBD::AioCompletion *)c;
-
-  // convert the scatter list into a bufferlist
-  auto aio_completion = get_aio_completion(comp);
-  ssize_t len = 0;
-  bufferlist bl;
+  len = 0;
   for (int i = 0; i < iovcnt; ++i) {
     const struct iovec &io = iov[i];
-    len += io.iov_len;
-    if (len < 0) {
-      break;
+    ssize_t len_tmp = len + io.iov_len;
```

Comment:
Why `ssize_t` when both `create_write_raw` and `aio_write` take an unsigned?

This is confusing because you are adding a signed (`len`) with an unsigned (`io.iov_len`) and then casting back to signed.  Let's stick to unsigned and do something that is easier to follow, e.g.:
```
size_t len = 0;
for (int i = 0; i < iovcnt; ++i) {
  const struct iovec &io = iov[i];
  if (len + io.iov_len < len) {
    return -EINVAL;
  }
  len += io.iov_len;
}
```

Reply:
My first version was using unsigned `size_t` and I was confused why the old code used `ssize_t` too,  but as far as I understand it the `rbd_aio_get_return_value` returns a `ssize_t` so aio functions can not read or write more than ssize_t positive max. This is implicitly taken care of by using ssize_t here.

Reply:
I think we can ignore that here as both `rbd_aio_write` and `rbd_aio_write2` have the same issue: their `len` is `size_t`.

Reply:
Ok so we use `size_t` here and add an additional check in aio_read because it returns byte read through `rbd_aio_get_return_value`. So this would change the semantic of aio_write* compared to before if that is ok I can change it accordingly.

Reply:
I'd not worry about `rbd_aio_get_return_value` here.  This PR is about iovec length computation overflow.  What `rbd_aio_get_return_value` does is more or less irrelevant because the non-v variants of both read and write take `size_t` for `len`.  If `get_iovec_length` computes `len` as unsigned and returns `size_t`, we become consistent with non-v variants which is a good thing.

Reply:
Ack. Will change it.